### ISSUE TITLE
TF2.0 fails post-training uint8 quantization

### ISSUE THREAD

**Issue by dojipkim** [tensorflow/tensorflow#35319]
**System information**
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Windows 10
- TensorFlow installed from (source or binary): installed TF binary 2.0 with Conda
- TensorFlow version (or github SHA if from source): TF 2.0

**Command used to run the converter or code if you’re using the Python API**

```
converter = tf.lite.TFLiteConverter.from_saved_model('model_mnist.hd5')
converter.optimizations = [tf.lite.Optimize.DEFAULT]
converter.target_spec.supported_ops = [tf.lite.OpsSet.TFLITE_BUILTINS_INT8]
converter.inference_input_type = tf.uint8
converter.inference_output_type = tf.uint8

images = tf.cast(X_train, tf.float32)
mnist_ds = tf.data.Dataset.from_tensor_slices((images)).batch(1)
def representative_data_gen():
    for input_value in mnist_ds.take(100):
        yield[input_value]
converter.representative_dataset = representative_data_gen

tflite_quant_model = converter.convert()
with open('model_mnist_quant_uint8.tflite', 'wb') as f:
    f.write(tflite_quant_model)


interpreter = tf.lite.Interpreter(model_path='model_mnist_quant_uint8.tflite')
interpreter.allocate_tensors()

img = X_train[0] * 255
img = img.astype('uint8')
print(interpreter.get_input_details())
interpreter.set_tensor(interpreter.get_input_details()[0]['index'], np.expand_dims(img, axis=0))
```

**The output from the converter invocation**

```
[{'name': 'flatten_input', 'index': 11, 'shape': array([ 1, 28, 28]), 'dtype': <class 'numpy.float32'>, 'quantization': (0.0, 0)}]
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-12-5a2fa86c9de2> in <module>
      5 img = img.astype('uint8')
      6 print(interpreter.get_input_details())
----> 7 interpreter.set_tensor(interpreter.get_input_details()[0]['index'], np.expand_dims(img, axis=0))

~\AppData\Local\Continuum\anaconda3\envs\tensorflow2\lib\site-packages\tensorflow_core\lite\python\interpreter.py in set_tensor(self, tensor_index, value)
    344       ValueError: If the interpreter could not set the tensor.
    345     """
--> 346     self._interpreter.SetTensor(tensor_index, value)
    347 
    348   def resize_tensor_input(self, input_index, tensor_size):

~\AppData\Local\Continuum\anaconda3\envs\tensorflow2\lib\site-packages\tensorflow_core\lite\python\interpreter_wrapper\tensorflow_wrap_interpreter_wrapper.py in SetTensor(self, i, value)
    134 
    135     def SetTensor(self, i, value):
--> 136         return _tensorflow_wrap_interpreter_wrapper.InterpreterWrapper_SetTensor(self, i, value)
    137 
    138     def GetTensor(self, i):

ValueError: Cannot set tensor: Got tensor of type UINT8 but expected type FLOAT32 for input 11, name: flatten_input 
```

**Also, please include a link to the saved model or GraphDef**

```
# Put link here or attach to the issue.
```

**Failure details**
The conversion is successful, but the generated model is wrong. The input tensor dtype should be uint8, but is still float32.

I tried the same thing with TF 1.15.0. In this case, every things works as expected. Here is the result with TF 1.15.0
```
[{'name': 'flatten_input', 'index': 11, 'shape': array([ 1, 28, 28], dtype=int32), 'dtype': <class 'numpy.uint8'>, 'quantization': (0.003921568859368563, 0)}]
```

**Any other info / logs**

Include any logs or source code that would be helpful to diagnose the problem. If including tracebacks, please include the full traceback. Large logs and files should be attached.

**Comment by jvishnuvardhan** [comment]
@dojipkim Can you please provide a standalone code to reproduce the issue? Thanks!

**Comment by dojipkim** [comment]
@jvishnuvardhan Yes. Here is the code.
```
import tensorflow as tf
from tensorflow import keras
from tensorflow.keras.models import Sequential
from tensorflow.keras import layers
import numpy as np

from tensorflow.keras.datasets import mnist
(X_train, y_train), (X_test, y_test) = mnist.load_data()
X_train = X_train.astype('float32') / 255.
X_test = X_test.astype('float32') / 255.
y_train = keras.utils.to_categorical(y_train)
y_test = keras.utils.to_categorical(y_test)

model = Sequential()
model.add(layers.Flatten(input_shape=X_train.shape[1:]))
model.add(layers.Dense(512, activation='relu'))
model.add(layers.Dense(10, activation='softmax'))
model.summary()

model.compile(optimizer='RMSprop', loss='categorical_crossentropy', metrics=['acc'])
history = model.fit(X_train, y_train, epochs=2, batch_size=32, validation_split=0.2)

model.save('model_mnist')

converter = tf.lite.TFLiteConverter.from_saved_model('model_mnist')
converter.optimizations = [tf.lite.Optimize.DEFAULT]
converter.target_spec.supported_ops = [tf.lite.OpsSet.TFLITE_BUILTINS_INT8]
converter.inference_input_type = tf.uint8
converter.inference_output_type = tf.uint8

images = tf.cast(X_train, tf.float32)
mnist_ds = tf.data.Dataset.from_tensor_slices((images)).batch(1)
def representative_data_gen():
    for input_value in mnist_ds.take(100):
        yield[input_value]
converter.representative_dataset = representative_data_gen

tflite_quant_model = converter.convert()
with open('model_mnist_quant_uint8.tflite', 'wb') as f:
    f.write(tflite_quant_model)

interpreter = tf.lite.Interpreter(model_path='model_mnist_quant_uint8.tflite')
interpreter.allocate_tensors()

img = X_train[0] * 255
img = img.astype('uint8')
print(interpreter.get_input_details())
interpreter.set_tensor(interpreter.get_input_details()[0]['index'], np.expand_dims(img, axis=0))
```

Thanks.

**Comment by amahendrakar** [comment]
@dojipkim,
Looks like the issue has been resolved with TensorFlow v2.3. I was able to run the code the code without any error. Please find the gist of it [here](https://colab.research.google.com/gist/amahendrakar/0651dda764b54bf618dfc380c7fb3f3c/35319.ipynb). Thanks!
